Treat unencrypted local origins as insecure context in Safari

jonkoops opened the pull request:
Closes #33557

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

jonkoops committed:
Treat unencrypted origins as insecure context in Safari

Closes #33557

Signed-off-by: Jon Koops <jonkoops@gmail.com>

douglaspalmer reviewed (approved):
This isn't compiling right now; I guess you missed some usages of isSecureContext() when you refactored. Looks good otherwise though.

douglaspalmer reviewed (changes requested):
Upon reflection I think there is a simpler way to do this. Instead of refactoring isSecureContext() and parsing the User Agent string it would be simpler and more efficient to simply use: 

String userAgent = context.getRequestHeaders().getHeaderString(HttpHeaders.USER_AGENT);
if(userAgent != null && userAgent.toLowerCase().contains("safari")) {
    return true;
}

shawkins commented on services/src/main/java/org/keycloak/utils/SecureContextResolver.java:
```diff
@@ -1,4 +1,7 @@
-package org.keycloak.common.util;
```
Just want to double check on how public are keycloak common apis? Do we have to give any consideration to backwards compatibility / deprecation in general?

stianst replied:
This file was moved here by mistake previously, and should not be part of common.

jonkoops replied:
Also considering this work will be in the upcoming patch release I don't think it will be too much of an issue.

stianst reviewed (changes requested):
Needs testing, see https://github.com/keycloak/keycloak/blob/main/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/cookies/DefaultCookieProviderTest.java. Should be relatively easy to send a http request with the Safari user-agent set. See some of the existing tests on how to do that.

One additional issue here is that this can lead to parsing the user-agent header multiple times for a single request. That can easily be fixed by updating https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/device/DeviceRepresentationProviderImpl.java to introduce a field `private DeviceRepresentation deviceRepresentation`, and only parse the header if this is not set.

stianst commented:
> Upon reflection I think there is a simpler way to do this. Instead of refactoring isSecureContext() and parsing the User Agent string it would be simpler and more efficient to simply use:
> 
> String userAgent = context.getRequestHeaders().getHeaderString(HttpHeaders.USER_AGENT); if(userAgent != null && userAgent.toLowerCase().contains("safari")) { return true; }

Makes more sense to me to do this properly and use the `DeviceRepresentationProvider` as we shouldn't introduce multiple places that parses user-agent headers.

jonkoops commented:
> `userAgent.toLowerCase().contains("safari")`

The issue with this code is that all browsers provide various user-agents other than their own in their user-agent string for backwards compatibility. This might lead to a browser being incorrectly identified, for example Google Chrome has the following user-agent:

`Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36`

As you can see it advertises itself as `Mozilla`, `AppleWebkit`, `Chrome` and `Safari`. So using a simple `contains("safari")` would lead to a false positive here.

jonkoops commented:
I've added unit and integration tests for the Safari quirks. I've also changed the implementation of the `SecureContextResolver` to take a `Supplier` so that the user agent can be evaluated lazily and properly unit tested without having to deal with mocking an implementation of `KeycloakSession`. This enures that on a production deployment (HTTPS) the user agent is never evaluated, so it should incur no performance hits.

shawkins reviewed (approved):
LGTM, I'll adjust #33530 as necessary if it moves forward.

douglaspalmer reviewed (approved):
LGTM
